src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -169,24 +220,28 @@ SHAMapInnerNode::makeCompressedInner(Slice data)
 
     SerialIter si(data);
 
-    auto ret = std::make_shared<SHAMapInnerNode>(0, branchFactor);
+    std::uint8_t nonEmptyBranches = data.size() / chunkSize;
+    auto ret = std::make_shared<SHAMapInnerNode>(0, nonEmptyBranches);
 
     auto hashes = ret->hashesAndChildren_.getHashes();
+    int prevPos = 0;
 
     while (!si.empty())
     {
         auto const hash = si.getBitString<256>();
         auto const pos = si.get8();
 
-        if (pos >= branchFactor)
-            Throw<std::runtime_error>("invalid CI node");
+        ret->isBranch_ |= (1 << pos);
+        auto const ix = ret->getChildIndex(pos);
 
-        hashes[pos].as_uint256() = hash;
+        if (pos >= branchFactor || prevPos > pos || !ix.has_value())
+            Throw<std::runtime_error>("invalid CI node");
 
-        if (hashes[pos].isNonZero())
-            ret->isBranch_ |= (1 << pos);
+        hashes[ix.value()].as_uint256() = hash;
+        prevPos = pos;
     }
 
+    // Should effectively be be a noop at this point
```

Comment:
so .... just remove it??

Reply:
```C++
void
SHAMapInnerNode::resizeChildArrays(std::uint8_t toAllocate)
{
    hashesAndChildren_ =
        TaggedPointer(std::move(hashesAndChildren_), isBranch_, toAllocate);
}
```

```C++
inline TaggedPointer::TaggedPointer(
    TaggedPointer&& other,
    std::uint16_t isBranch,
    std::uint8_t toAllocate)
    : TaggedPointer(std::move(other))
{
    auto const oldNumAllocated = capacity();
    toAllocate = numAllocatedChildren(toAllocate);
    if (toAllocate == oldNumAllocated)
        return;
```

```C++
inline TaggedPointer::TaggedPointer(TaggedPointer&& other) : tp_{other.tp_}
{
    other.tp_ = 0;
}
```


It seems it would still be a new pointer, but a redundant one.

I removed the call to `resizeChildArrays` in `SHAMapInnerNode::makeFullInner` but was inconsistent, forgetting to apply it `SHAMapInnerNode::makeCompressedInner`.

I just pushed a commit to consistently remove it.